### PR TITLE
feat(elec): ground service buses

### PR DESCRIPTION
**Pull request by davidwalschots**
## Summary of Changes
This PR adds the AC GND/FLT and DC GND/FLT buses to support various ground service related scenarios such as refueling. It builds on top of the work done in #2582.

## Functionality

@komp1821 has the following working theory regarding ground service buses for now:

> Whenever AC BUS 2 is available, the ground service network is supplied by AC BUS 2 and DC BUS 2. 
> If AC BUS 2 is not available, supply is from External power, if available.

The above means that DC GND/FLT will never be supplied when neither AC BUS 2 nor EXT PWR is available.

This results in the following functionality:

- When AC BUS 2 is powered, and TR2 hasn't failed:
    - [x] Contactor 12XN is open.
    - [x] Contactor 3PX is open.
    - [x] Contactor 14PU is closed.
    - [x] Contactor 5PU2 is closed.
    - [x] Contactor 8PN is closed
    - [x] AC GND/FLT is powered.
    - [x] DC GND/FLT is powered.
- When AC BUS 2 is unpowered, and EXT PWR is AVAIL:
    - [x] Contactor 12XN is closed.
    - [x] Contactor 3PX is closed.
    - [x] Contactor 14PU is open.
    - [x] Contactor 5PU2 is open.
    - [x] Contactor 8PN is open.
    - [x] AC GND/FLT is powered.
    - [x] DC GND/FLT is powered.
- When AC BUS 2 is powered, and TR2 has failed:
    - [x] Contactor 12XN is open.
    - [x] Contactor 3PX is open.
    - [x] Contactor 14PU is open.
    - [x] Contactor 5PU2 is open.
    - [x] Contactor 8PN is open.
    - [x] AC GND/FLT is unpowered.
    - [x] DC GND/FLT is unpowered.
- When AC BUS 2 is unpowered, EXT PWR is AVAIL, and TR2 has failed:
    - [x] Contactor 12XN is open.
    - [x] Contactor 3PX is open.
    - [x] Contactor 14PU is open.
    - [x] Contactor 5PU2 is open.
    - [x] Contactor 8PN is open.
    - [x] AC GND/FLT is unpowered.
    - [x] DC GND/FLT is unpowered.

## References
@komp1821 

## Additional context

Discord username (if different from GitHub): SjotgunSjonnie#9623

## Testing instructions
As nothing currently uses the ground service buses, there isn't really a visible way to verify correctness.
One could use the Dev Mode and check the local variables (Windows -> Model Behaviors -> LocalVariables tab):

- A32NX_ELEC_AC_GND_FLT_SVC_BUS_IS_POWERED
- A32NX_ELEC_DC_GND_FLT_SVC_BUS_IS_POWERED
- A32NX_ELEC_CONTACTOR_14PU_IS_CLOSED
- A32NX_ELEC_CONTACTOR_12XN_IS_CLOSED
- A32NX_ELEC_CONTACTOR_3PX_IS_CLOSED
- A32NX_ELEC_CONTACTOR_8PN_IS_CLOSED

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

### CODE REVIEW

**Comment by avteknisyan** [comment]
> Whenever AC BUS 2 is available, the ground service network is supplied by AC BUS 2 and DC BUS 2.
> If AC BUS 2 is not available, supply is from External power, if available.

this is correct.

when GPU is connected and power is ok.
on the external panel the GPU White "Not in Use" and Amber "Avail" Light is on

you can supply the AC GND/FLT & DC GND/FLT (through TR2) Buses by selecting the MAINT BUS SW to On

When you Switch the EXT PWR Pushbutton switch on the overhead panel, 

DC GND/FLT & AC/GND FLT Buses will be supplied through AC Bus 2 (for AC) and TR2 (for DC)

**Comment by davidwalschots** [comment]
@avteknisyan The biggest doubt is in whether or not DC GND/FLT gets powered through AC1 when AC2 fails (resulting in AC GND/FLT not being powered) in a non-EXT PWR scenario. So for example GEN1 supplying power through TR1 to DC GND/FLT.

**Comment by avteknisyan** [comment]
> @avteknisyan The biggest doubt is in whether or not DC GND/FLT gets powered through AC1 when AC2 fails (resulting in AC GND/FLT not being powered) in a non-EXT PWR scenario. So for example GEN1 supplying power through TR1 to DC GND/FLT.

So scenario is, Loss of AC Bus 2 with GEN 1 operating and GEN 2 inoperative?

**Comment by davidwalschots** [comment]
@avteknisyan That's an example scenario indeed. Though I guess GEN1, GEN2, or GEN1+2 shouldn't matter. The main question is whether DC GND/FLT is powered through the AC1 -> TR1 -> DC1 -> DC BAT -> DC2 -> DC GND/FLT route.

**Comment by avteknisyan** [comment]
> @avteknisyan That's an example scenario indeed. Though I guess GEN1, GEN2, or GEN1+2 shouldn't matter. The main question is whether DC GND/FLT is powered through the AC1 -> TR1 -> DC1 -> DC BAT -> DC2 -> DC GND/FLT route.

yes it will supplied due to loss of TR2 which is consequence of AC Bus 2 Loss

DC Bus 2 (2PP) >> DC SVC 6PP Supply Ckt Breaker (7PN) >> DC Svc Bus Normal Supply Contactor (8PN)  >> DC Service Bus (6PP)

DC


Contactor 3PX is only closed when on Ground Supply

**Comment by komp1821** [comment]
Sorry, but just because there is a possible path in the schematic doesn't mean that this switching combination is actually possible. In this case, it is not possible. There are a few interlocking conditions.

In short: 8PN is in series with 14PU, which is closed when 7XX is energized, which is, guess what, only energized when AC BUS 2 is available.

**Comment by avteknisyan** [comment]
> Sorry, but just because there is a possible path in the schematic doesn't mean that this switching combination is actually possible. In this case, it is not possible. There are a few interlocking conditions not shown in the general schematic.
> 
> In short: 8PN is in series with 14PU, which is closed when 7XX is energized, which is, guess what, only energized when AC BUS 2 is available. Remember that these relays have multiple switching layers.

oh. i've overlooked the the other relays, thanks for correcting!

File: src/systems/a320_systems/src/electrical/alternating_current.rs [hunk 1]
```diff
@@ -81,8 +89,28 @@ impl A320AlternatingCurrentElectrical {
...
+        self.ac_bus_2_to_tr_2_contactor.powered_by(&self.ac_bus_2);
+        self.ac_bus_2_to_tr_2_contactor
+            .close_when(self.ac_bus_2.is_powered() && !self.tr_2.failed());
+
+        self.ext_pwr_to_ac_gnd_flt_service_bus_and_tr_2_contactor
+            .powered_by(ext_pwr);
+
+        // On the real aircraft there is a button inside the galley which is taken into
+        // account when determining whether to close this contactor or not.
+        // As we're not building a galley simulator, for now we assume the button is ON.
```
Comment (Benjozork) [inline src/systems/a320_systems/src/electrical/alternating_current.rs:101]: Don't give us ideas 😆

Comment (davidwalschots) [reply src/systems/a320_systems/src/electrical/alternating_current.rs:101]: 🤣

Comment (MisterChocker) [reply src/systems/a320_systems/src/electrical/alternating_current.rs:101]: I'd like a galley simulator, FBW now also offers fully cabin crew simulation xD ahahahah.

**Review by MisterChocker — APPROVED** [review]
LGTM from a syntax POV, altho I am generally confused by the rust formatting xD

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 2.05.2021
Version of the sim: 1.15.8.0
PR Tested: #4628
Tier of testing: 1
Changes to observe: GEN 1 LINE
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. 
3.) When switch on ext. power:
- A32NX_ELEC_AC_GND_FLT_SVC_BUS_IS_POWERED 1>1
- A32NX_ELEC_DC_GND_FLT_SVC_BUS_IS_POWERED 1>1
- 14PU_IS_CLOSED 0>1
- 12XN_IS_CLOSED 1>0
- 3PX_IS_CLOSED 1>0
- 8PN_IS_CLOSED 0>1
- 5PU2_IS_CLOSED 0>1

3.) Pushback off parking position (ext. power not avail.)
- A32NX_ELEC_AC_GND_FLT_SVC_BUS_IS_POWERED 0
- A32NX_ELEC_DC_GND_FLT_SVC_BUS_IS_POWERED 0
- 14PU_IS_CLOSED 0
- 12XN_IS_CLOSED 0
- 3PX_IS_CLOSED 0
- 8PN_IS_CLOSED 0
- 5PU2_IS_CLOSED 0

4.) AC BUS 2 is powered and TR2 hasn't failed
- A32NX_ELEC_AC_GND_FLT_SVC_BUS_IS_POWERED 1
- A32NX_ELEC_DC_GND_FLT_SVC_BUS_IS_POWERED 1
- 12XN_IS_CLOSED 0
- 3PX_IS_CLOSED 0
- 14PU_IS_CLOSED 1
- 5PU2_IS_CLOSED 1
- 8PN_IS_CLOSED 1

**Don't know how to put TR2 in fail**

Issues: -

Overall Rating: Good

Conclusions of the testing: Tested the parts without TR2 fail. States are like in the description.

Hope this helps :-)

**Comment by davidwalschots** [comment]
@marcman86 thank you for testing. It's indeed not yet possible to test/simulate TR2 failure in the sim at this point.

**Pull request merged**